### PR TITLE
gh-63284: Add support for TLS-PSK (pre-shared key) to the ssl module

### PR DESCRIPTION
**Pull request by grantramsay**
Add support for TLS-PSK (pre-shared key) to the ssl module (plus documentation and unit-tests).

The referenced issue (https://github.com/python/cpython/issues/63284) is ~10 years old, but the information is still valid.

I've tested this by:

- connecting python TLS-PSK client/server (as in the unit-tests)
- connecting to openssl s_client/s_server from python:
  - `openssl s_server -accept localhost:12345 -psk deadbeef -cipher PSK -verify_return_error -nocert`
  - `openssl s_client -connect localhost:12345 -psk deadbeef -cipher PSK -verify_return_error`
  
Things I'm still uncertain on:

- ~~It's not yet working with TLS 1.3 (I'm currently looking into this).~~ [Fixed]
- ~~Do I need to call `PyErr_WriteUnraisable()` if there's a python exception during a C callback?~~ [Fixed]

This is my first python contribution.
I have probably got a few things wrong, but I'm happy to fix them up.

Thanks!

* Issue: gh-63284

### CODE REVIEW

**Comment by grantramsay** [comment]
I figured out why it wasn't working with TLS 1.3.

The error on the client-side is:

    ssl.SSLError: [SSL: ATTEMPT_TO_REUSE_SESSION_IN_DIFFERENT_CONTEXT] attempt to reuse session in different context (_ssl.c:996)

When initialising the SSLContext there is a call to `SSL_CTX_set_session_id_context()`:

	#define SID_CTX "Python"
	    SSL_CTX_set_session_id_context(self->ctx, (const unsigned char *) SID_CTX,
	                                   sizeof(SID_CTX));
	#undef SID_CTX

The openssl man pages state this is a "server side only" operation:
https://www.openssl.org/docs/man1.0.2/man3/SSL_CTX_set_session_id_context.html

> SSL_CTX_set_session_id_context, SSL_set_session_id_context - set context within which session can be reused (server side only)

> The session id context becomes part of the session. The session id context is set by the SSL/TLS server. The SSL_CTX_set_session_id_context() and SSL_set_session_id_context() functions are therefore only useful on the server side.

Calling this on a client-side socket seems to result in unexpected behavior, where the client thinks it's trying to resume a non-existent session.

It looks like others have run into the same issue: https://github.com/maovidal/paho_sslpsk2_demo/issues/2#issuecomment-1233488678
	
Disabling the call to `SSL_CTX_set_session_id_context()` for client-side SSLContext creation fixes it.
I've added another commit with this fix

**Comment by arhadthedev** [comment]
@jackjansen, @dstufft, @alex (as ssl module [experts](https://devguide.python.org/core-developers/experts/))

**Comment by grantramsay** [comment]
Modified the documentation for TLS 1.3.
I initially assumed it was a quirk of the OpenSSL TLS 1.3 implementation that meant "client-identity" had to be a non-empty string.

However this ticket explains that it is a change in TLS 1.3: https://github.com/openssl/openssl/issues/8894#issuecomment-490560793

> TLSv1.3 explicitly states that an identity must be at least 1 byte long. But in TLSv1.2 an identity was allowed to have a zero length.

**Comment by grantramsay** [comment]
Hey just giving this a bump

File: Modules/_ssl.c [hunk 1]
```diff
@@ -3167,10 +3171,14 @@ _ssl__SSLContext_impl(PyTypeObject *type, int proto_version)
        usage for no cost at all. */
     SSL_CTX_set_mode(self->ctx, SSL_MODE_RELEASE_BUFFERS);
 
+    /* Setting the session id context is a server-side only operation.
```
Comment (gpshead) [inline Modules/_ssl.c]: Can you make a separate PR for this little logic change about only calling set_session_id_context for servers?  It seems like a bugfix that can stand on its own.

File: Modules/_ssl.c [hunk 2]
```diff
@@ -4619,6 +4627,202 @@ _ssl__SSLContext_get_ca_certs_impl(PySSLContext *self, int binary_form)
...
+    if (ssl == NULL || ssl->ctx == NULL) {
+        goto error;
+    }
+    PyObject *callback = ssl->ctx->psk_client_callback;
+    if (callback == NULL) {
+        goto error;
+    }
+
+    PyObject *hint_str = (hint != NULL) ?
+            PyUnicode_DecodeASCII(hint, strlen(hint), "strict") :
```
Comment (gpshead) [inline Modules/_ssl.c]: Why ASCII? I suspect "hint" should really be defined as PyBytes rather than PyUnicode.

File: Doc/library/ssl.rst [hunk 3]
```diff
@@ -1987,6 +1987,89 @@ to speed up repeated connections from the same clients.
...
          <VerifyMode.CERT_REQUIRED: 2>
 
+.. method:: SSLContext.set_psk_client_callback(callback)
+
+   Enables TLS-PSK (pre-shared key) authentication on a client-side connection.
+
+   In general, certificate based authentication should be preferred over this method.
+
+   The parameter ``callback`` is a callable object with the signature:
+   ``def callback(hint: str | None) -> tuple[str | None, bytes]``.
```
Comment (gpshead) [inline Doc/library/ssl.rst:2016]: per comment in the code, hint should probably be `bytes | None`

**Review by gpshead — CHANGES_REQUESTED** [review]

File: Doc/library/ssl.rst [hunk 4]
```diff
@@ -1987,6 +1987,89 @@ to speed up repeated connections from the same clients.
...
+
+      # A table using the hint from the server:
+      psk_table = { 'ServerId_1': bytes.fromhex('deadbeef'),
+                    'ServerId_2': bytes.fromhex('cafebabe')
+      }
+      def callback(hint):
+          return 'ClientId_1', psk_table[hint]
+      context.set_psk_client_callback(callback)
+
+   .. versionadded:: 3.12
```
Comment (gpshead) [inline Doc/library/ssl.rst]: The versionadded fields will need to be 3.13, we've missed the 3.12 feature freeze.

File: Modules/_ssl.c (hunk 2 above)
Comment (gramsay0) [reply Modules/_ssl.c]: I changed this to UTF-8 as specified in the RFC and added a test with non-ASCII chars.
https://www.rfc-editor.org/rfc/rfc4279

> 5.1.  PSK Identity Encoding
The PSK identity MUST be first converted to a character string, and then encoded to octets using UTF-8.

File: Doc/library/ssl.rst (hunk 3 above)
Comment (gramsay0) [reply Doc/library/ssl.rst:2016]: Changed to UTF-8 so keeping this as a string

File: Modules/_ssl.c (hunk 1 above)
Comment (grantramsay) [reply Modules/_ssl.c]: Issue: https://github.com/python/cpython/issues/105293
PR: https://github.com/python/cpython/pull/105295

**Comment by grantramsay** [comment]
I have made the requested changes; please review again

File: Modules/_ssl.c (hunk 1 above)
Comment (gpshead) [reply Modules/_ssl.c]: I'm going to leave this comment unresolved as a reminder that the other PR needs to be merged first.

**Review by gpshead — APPROVED** [review]

File: Modules/_ssl.c [hunk 5]
```diff
@@ -4641,6 +4645,202 @@ _ssl__SSLContext_get_ca_certs_impl(PySSLContext *self, int binary_form)
...
+    if (ssl == NULL || ssl->ctx == NULL) {
+        goto error;
+    }
+    PyObject *callback = ssl->ctx->psk_client_callback;
+    if (callback == NULL) {
+        goto error;
+    }
+
+    PyObject *hint_str = (hint != NULL) ?
+            PyUnicode_DecodeUTF8(hint, strlen(hint), "strict") :
```
Comment (gpshead) [inline Modules/_ssl.c:4691]: how safe is it to call `strlen(hint)`? this is raw data being supplied by OpenSSL, does it guarantee a reasonable length and null termination for us? (no hint length is passed, so i'm assuming so)

should the hint be treated as UTF-8?  Does TLS or OpenSSL require that?  It might make more sense to have code use PyBytes instead.  A failed UTF-8 decode is something the other end of the network connection could trigger.  same for the other callback regarding identity.

File: Modules/_ssl.c [hunk 6]
```diff
@@ -4641,6 +4645,202 @@ _ssl__SSLContext_get_ca_certs_impl(PySSLContext *self, int binary_form)
...
+    memcpy(psk, psk_, psk_len_);
+    memcpy(identity, identity_, identity_len_);
+    identity[identity_len_] = 0;
+
+    Py_DECREF(result);
+
+    PyGILState_Release(gstate);
+    return (unsigned int)psk_len_;
+
+error:
```
Comment (gpshead) [inline Modules/_ssl.c:4731]: Your thought in the PR text sounds correct: doing

```c
if (PyErr_Occurred()) {
    PyErr_WriteUnraiseable(callback);
    PyErr_Clear();
}
```

makes sense here before releasing the GIL.  same in the other callback.

File: Modules/_ssl.c [hunk 7]
```diff
@@ -4641,6 +4645,202 @@ _ssl__SSLContext_get_ca_certs_impl(PySSLContext *self, int binary_form)
...
+    PySSLSocket *ssl = SSL_get_app_data(s);
+    if (ssl == NULL || ssl->ctx == NULL) {
+        goto error;
+    }
+    PyObject *callback = ssl->ctx->psk_server_callback;
+    if (callback == NULL) {
+        goto error;
+    }
+
+    PyObject *identity_str = (identity[0] != '\0') ?
```
Comment (gpshead) [inline Modules/_ssl.c]: coding defensively, include `identity != NULL || ...` in that condition.

File: Modules/_ssl.c [hunk 8]
```diff
@@ -4641,6 +4645,202 @@ _ssl__SSLContext_get_ca_certs_impl(PySSLContext *self, int binary_form)
...
+            PyUnicode_DecodeUTF8(identity, strlen(identity), "strict") :
```
Comment (gpshead) [inline Modules/_ssl.c:4797]: why utf-8 rather than bytes?

File: Modules/_ssl.c [hunk 9]
```diff
@@ -4641,6 +4645,202 @@ _ssl__SSLContext_get_ca_certs_impl(PySSLContext *self, int binary_form)
...
+    PyObject *result = PyObject_CallFunctionObjArgs(callback, identity_str, NULL);
+    Py_DECREF(identity_str);
+
+    if (result == NULL) {
+        goto error;
+    }
+
+    const char *psk_;
+    Py_ssize_t psk_len_;
+    if (!PyArg_Parse(result, "y#", &psk_, &psk_len_)) {
```
Comment (gpshead) [inline Modules/_ssl.c]: rather than using `PyArg_Parse` here, use:

if (PyBytes_AsStringAndSize(&psk_, &psk_len_) < 0) {
}

it's a single value, there is no need to parse anything.

**Review by gpshead — CHANGES_REQUESTED** [review]

File: Modules/_ssl.c (hunk 5 above)
Comment (gramsay0) [reply Modules/_ssl.c:4691]: > does it guarantee a reasonable length and null termination for us?

Yes:

> The callback function is given the connection in parameter ssl, a NULL-terminated PSK identity hint sent by the server in parameter hint

The OpenSSL implementation limits `identity` and `identity_hint` string length to `#define PSK_MAX_IDENTITY_LEN 256`, and uses `strndup` to guarantee null termination.
(Oddly this doesn't seem to meet identity length specified in the RFC `opaque psk_identity<0..2^16-1>;`)

> should the hint be treated as UTF-8? Does TLS or OpenSSL require that?

The RFC requires `identity` and `identity_hint` to be encoded as UTF-8:

> 5.1. PSK Identity Encoding
The PSK identity MUST be first converted to a character string, and then encoded to octets using UTF-8.

Although, the OpenSSL implementation doesn't do any UTF-8 validation, it treats what it is given as null terminated bytes. So it would be possible to cause a python exception on the remote side by sending invalid UTF-8 (I assume that is what you would like to avoid).

I think a good approach would be for the C callbacks to just ignore/drop requests with invalid UTF-8. Although this would require a UTF-8 validation function in C that does not raise a python exception (not sure if that exists).

For now I will follow your advice and change these to PyBytes

File: Modules/_ssl.c [hunk 10]
```diff
@@ -3116,6 +3118,8 @@ _ssl__SSLContext_impl(PyTypeObject *type, int proto_version)
     self->alpn_protocols = NULL;
     self->set_sni_cb = NULL;
     self->state = get_ssl_state(module);
+    self->psk_client_callback = NULL;
+    self->psk_server_callback = NULL;
```
Comment (gramsay0) [inline Modules/_ssl.c:3129]: Free callbacks when SSL context is deallocated

File: Modules/_ssl.c [hunk 11]
```diff
@@ -4641,6 +4645,202 @@ _ssl__SSLContext_get_ca_certs_impl(PySSLContext *self, int binary_form)
...
+        Py_INCREF(callback);
+        ssl_callback = psk_server_callback;
+    }
+
+    if (self->psk_server_callback != NULL) {
+        Py_DECREF(self->psk_server_callback);
+    }
+    self->psk_server_callback = callback;
+    SSL_CTX_set_psk_server_callback(self->ctx, ssl_callback);
+    SSL_CTX_use_psk_identity_hint(self->ctx, identity_hint);
```
Comment (gramsay0) [inline Modules/_ssl.c]: Check return value of `SSL_CTX_use_psk_identity_hint`

File: Doc/library/ssl.rst [hunk 12]
```diff
@@ -1987,6 +1987,89 @@ to speed up repeated connections from the same clients.
          >>> ssl.create_default_context().verify_mode  # doctest: +SKIP
          <VerifyMode.CERT_REQUIRED: 2>
 
+.. method:: SSLContext.set_psk_client_callback(callback)
```
Comment (gramsay0) [inline Doc/library/ssl.rst:2009]: Add documentation on:

- the length limits of the `identity` and `hint` parameters
- how to reject a connection (return a PSK of zero length)

File: Modules/_ssl.c [hunk 13]
```diff
@@ -4641,6 +4645,202 @@ _ssl__SSLContext_get_ca_certs_impl(PySSLContext *self, int binary_form)
...
+        if (!PyCallable_Check(callback)) {
+            PyErr_SetString(PyExc_TypeError, "callback must be callable");
+            return NULL;
+        }
+        Py_INCREF(callback);
+        ssl_callback = psk_server_callback;
+    }
+
+    if (self->psk_server_callback != NULL) {
+        Py_DECREF(self->psk_server_callback);
```
Comment (gramsay0) [inline Modules/_ssl.c]: Make use of `Py_XINCREF` and `Py_XDECREF` for tidier NULL checking

**Review by gramsay0 — COMMENTED** [review]
Leaving myself notes on things to tidy up

File: Modules/_ssl.c (hunk 5 above)
Comment (gramsay0) [reply Modules/_ssl.c:4691]: > For now I will follow your advice and change these to PyBytes

@gpshead I've changed my mind on this.
I don't think it makes sense to use PyBytes to handle a UTF-8 string.
This could stay as UTF-8 and call `PyErr_Clear();` if the UTF-8 decode fails and reject the connection.

Could you please elaborate on your concerns if you think this should still be PyBytes?

**Comment by grantramsay** [comment]
I have made the requested changes; please review again

**Comment by bruno-at-bareos** [comment]
@gpshead is there anything that still need to be done to see this PR moving forward?

**Comment by grantramsay** [comment]
Hi @gpshead, it has been a while but I'm still interested in getting this merged.
Do you have any further review comments/suggestions?

I have made the requested changes; please review again

**Comment by doronz88** [comment]
I'm also very interested in this feature. Feel free to tag me if you need any help

File: Modules/_ssl.c (hunk 5 above)
Comment (gpshead) [reply Modules/_ssl.c:4691]: your reasoning makes sense and I like your PyErr_Clear approach when the Decode fails.

File: Modules/_ssl.c [hunk 14]
```diff
@@ -4662,6 +4668,220 @@ _ssl__SSLContext_get_ca_certs_impl(PySSLContext *self, int binary_form)
...
+        Py_DECREF(result);
+        goto error;
+    }
+
+    if (identity_len_ + 1 > max_identity_len || psk_len_ > max_psk_len) {
+        Py_DECREF(result);
+        goto error;
+    }
+    memcpy(psk, psk_, psk_len_);
+    memcpy(identity, identity_, identity_len_);
```
Comment (gpshead) [inline Modules/_ssl.c]: i added a check that `identity_` was not NULL here as the "z#" above can return NULL but doesn't specify if `identity_len_` will be set to anything in that situation.

**Review by gpshead — APPROVED** [review]

**Comment by gpshead** [comment]
Thanks for the contributon!

FWIW the timestamp on the NEWS.d/next/ filename is irrelevant, it is only being used as a uniqueness collision avoidance mechanism. All of those get merged into one file during releases.

**Comment by grantramsay** [comment]
Awesome!! Thanks for taking the time to review!
